res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1336 @@
+      } else {
+        script.triggerControl(group, "beatjump_backward");
+      }
+    }
+  }
+}
+
+TraktorS2MK2.leftEncoderPress = function(field) {
+  var splitted = field.id.split(".");
+  var group = splitted[0];
+  TraktorS2MK2.left_encoder_pressed[group] = (field.value > 0) ? true : false;
+  if (TraktorS2MK2.shift_pressed[group] && field.value > 0) {
+      script.triggerControl(group, "pitch_adjust_set_default");
+  }
+}
+
+TraktorS2MK2.rightEncoder = function(field) {
+  var splitted = field.id.split(".");
+  var group = splitted[0]
+  previous_rightencoder = TraktorS2MK2.previous_leftencoder[group];
+  TraktorS2MK2.previous_leftencoder[group] = field.value;
+  var delta = 0;
+  if (previous_rightencoder === 15 && field.value === 0) {
+    delta = 1;
+  } else if (previous_rightencoder === 0 && field.value === 15) {
+    delta = -1;
+  } else if (field.value > previous_rightencoder) {
+    delta = 1;
+  } else {
+    delta = -1;
+  }
+
+  if (TraktorS2MK2.shift_pressed[group]) {
+    if (delta == 1) {
+      script.triggerControl(group, "beatjump_1_forward");
+    } else {
+      script.triggerControl(group, "beatjump_1_backward");
+    }
+  } else {
+    if (delta == 1) {
+      script.triggerControl(group, "loop_double");
+    } else {
+      script.triggerControl(group, "loop_halve");
+    }
+  }
+}
+
+TraktorS2MK2.rightEncoderPress = function(field) {
+  if (field.value === 0) {
+      return;
+  }
+  var splitted = field.id.split(".");
+  var group = splitted[0];
+  var loop_enabled = engine.getValue(group, "loop_enabled");
+  // The actions triggered below change the state of loop_enabled,
+  // so to simplify the logic, use script.triggerControl to only act
+  // on press rather than resetting ControlObjects to 0 on release.
+  if (TraktorS2MK2.shift_pressed[group]) {
+    if (loop_enabled) {
+      script.triggerControl(group, "reloop_andstop");
+    } else {
+      script.triggerControl(group, "reloop_toggle");
+    }
+  } else {
+    if (loop_enabled) {
+      script.triggerControl(group, "reloop_toggle");
+    } else {
+      script.triggerControl(group, "beatloop_activate");
+    }
+  }
+}
+
+TraktorS2MK2.callbackBrowse = function(field) {
+  // TODO: common-hid-packet-parser looks like it should do deltas, but I can't get them to work.
+  prev_browse = TraktorS2MK2.previous_browse;
+  TraktorS2MK2.previous_browse = field.value;
+  var delta = 0;
+  if (prev_browse === 15 && field.value === 0) {
+    delta = 1;
+  } else if (prev_browse === 0 && field.value === 15) {
+    delta = -1;
+  } else if (field.value > prev_browse) {
+    delta = 1;
+  } else {
+    delta = -1;
+  }
+
+  engine.setValue("[Playlist]", "SelectTrackKnob", delta);
+}
+
+TraktorS2MK2.scalerParameter = function(group, name, value) {
+  return script.absoluteLin(value, 0, 1, 16, 4080);
+}
+// Tell the HIDController script to use setParameter instead of setValue.
+TraktorS2MK2.scalerParameter.useSetParameter = true;
```

Comment:
should this be at the top of the file?

Reply:
No, TraktorS2MK2.scalerParameter is defined right above this.

Reply:
Why is this done outside of the functions, though? it looks like it should happen in init()

Reply:
67fbe3d27dc97e6fc280c5626e8016a5f67a0e64